### PR TITLE
feat(docs):  make docs development practical

### PR DESCRIPTION
**Pull request by vwheeler63**
I have modified `build.py` as follows:

- I have been using this (and thus testing it) for about 10 weeks now and it has been stable for the last 6 weeks.

- All functionality is retained.

- Already-existing command line (as used in CI/CD tests) functions identically -- there is no change in output results.

- Logging output to STDOUT was enhanced to include more details of what is happening during doc generation -- content I found I needed to see, as well as generation time information.

- Functionality not needed in a docs-dev environment (where you want to be able to make a modification and then re-generate docs quickly over and over with a short turn-around time) was made so that each major feature can be individually disabled using command-line arguments.

- How to use all arguments is now fully documented in `build.py` module's docstring.

- Major code sections are now marked with block comments.

- `build.py` module was made to be able to be loaded with an `import` statement as well as run directly (this is considered a good coding practice in the Python world).  (Credit goes to @kdschlosser for indirectly making me aware this was a good idea.)

- 3 logging STDOUT outputs were added to `doc_builder.py`.

- 1 additional (well proven) Python package requirement (dirsync) was added and is used in the docs-dev environment to quickly update master-doc edits to the temporary (generation) directory before re-running `sphinx-build`.  This is part of about 6 different things that make it possible to dramatically shorten edit-to-generation turn-around time required when repeatedly generating the documentation to test output.  Speed statistics are documented in `build.py`.

Fixes to `doc_builder.py`:  when used from `gen_json.py`, it had some bugs causing to not run correctly on a Windows platform where TMPDIR environment variable set to a different drive than the script is executing from.  Specifically:

- when Python's `tempfile.mkdtemp()` is on a different drive from where the script was executing, it causes  the "cd tmpdir && doxygen Doxyfile" syntax to fail because of the different semantics of the `cd` command on Windows:  it doesn't change the default DRIVE if `cd` is executed from a different drive.  The result, when this is the case, is that Doxygen runs in the current working directory instead of in the temporary directory as was intended.  (fixed)
- when Doxygen exited with a non-zero result code, `sys.stdout.write()` was being called passing it a byte array instead of a string, and this caused it the Python script to generate an exception and abort.  (fixed)
- LVGL_URLPATH environment variable was being populated with a multi-line list of all current LVGL branches, rather than just the name of the current branch as intended.  (fixed)

The result:  the `gen_json.py` script now runs correctly on a Windows platform with TMPDIR environment variable set to a different drive than the script is executing from.

cc:  @kisvegabor @liamHowatt 

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.  (Done within the `build.py` module.)
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.  n/a
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.  n/a
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).  n/a
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).  n/a
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.  Done.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.  n/a
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

File: docs/doc_builder.py
Comment (kdschlosser) [inline docs/doc_builder.py]: as a suggestion I would go about handling these kinds of statements like so...


```python

def doc_print(*args):
    args = ' '.join(repr(arg) for arg in args)
    print(f'{os.path.basename(__file__)}:', args)  
```

and you use it like so...

```python
doc_print('Generating API documentation .RST files...')
```

Comment (kdschlosser) [reply docs/doc_builder.py]: Now one of the things you have to remember is the gen_json script which dumps the JSON data to STDOUT That script uses the documentation build system to work. so if you have all kinds of output happening it may cause an issue. If you notice the error output is guarded by a constant. This is done so the output is able to be turned off so it doesn't pollute the STDOUT when using the gen_json script.

Comment (vwheeler63) [reply docs/doc_builder.py]: Hi, Kevin!

> as a suggestion...

Thank you!  Nice logic.  I may throw that in later.

> Now one of the things you have to remember is the gen_json script which dumps the JSON data to STDOUT That script uses the documentation build system to work. so if you have all kinds of output happening it may cause an issue. If you notice the error output is guarded by a constant. This is done so the output is able to be turned off so it doesn't pollute the STDOUT when using the gen_json script.

Q1:  Having a look at this, I see that `./scripts/gen_json/pycparser_monkeypatch.py` is the only `.py` file in the `lvgl/lvgl` repository that mentions `doc_builder` (outside of `build.py`) and it calls `doc_builder.XMLSearch()`, which does not call the `doc_builder.py::run()` function, which is the only place I put `print()` calls.  So my modification does not appear to get called at any point from `gen_json.py`.  Only from `build.py` when it calls `doc_builder.run()`.  Would you concur?

Q2:  When does `gen_json.py` get used?  (I see it in `./tests/gen_json/test_gen_json.py`, and `.github/workflows/gen_json.yml`, and it is mentioned in `./docs/details/integration/bindings/api_json.rst` --where it appears to be described as a hand-run script to get API info.  But I am unclear on if that is the only place, or if it gets run as part of a test suite (CI?) or by hand when _____ ?)  I.e. what are the circumstances when this is or should be run?  It appears to be to help automate building LVGL bindings.  Is that the extent of its use (i.e. programmers running it by hand to generate API JSON data)?

Comment (kdschlosser) [reply docs/doc_builder.py]: I didn't go and check specifically if it would cause an issue or not. I knew that the gen_json script imported the doc_builder script. I was telling you so that you knew to be careful because it could break other things by adding print statements if they happen to end up in a location that would print from the json_gen script.


The gen_json script is used in the MicroPython Binding and I know there was work to use it in one other binding as well.

Comment (kdschlosser) [reply docs/doc_builder.py]: It can be programmatically run and the output captured or it can also be imported and used that way as well.

Comment (kdschlosser) [reply docs/doc_builder.py]: It does one other really important feature. It collects all the documentation from LVGL and makes it easy to access for say a binding's help.

The script has also been used to locate things that need to be changed in the API. It's very handy and can be used for a plethora of different things.

Comment (vwheeler63) [reply docs/doc_builder.py]: > I was telling you so that you knew to be careful because it could break other things by adding print statements if they happen to end up in a location that would print from the json_gen script.

Got it!  I was not aware that `gen_json.py` also used it, but am now, and after examining it closely, find that my `print()` calls are not at risk of interfering with `gen_json.py`.

I've observed what it does -- extracting content out of the Doxygen XML and generating an in-RAM data structure from it.  Agreed:  powerful!  :-)

**Comment by vwheeler63** [comment]
@liamHowatt @kisvegabor @kdschlosser 

In confirming Kevin's concerns above were taken care of, I ran `./scripts/gen_json/gen_json.py` which uses `doc_builder.py` and found some bugs in `doc_builder.py` and fixed them:

- when Python's `tempfile.mkdtemp()` is on a different drive from where the script was executing, it causes  the "cd tmpdir && doxygen Doxyfile" syntax to fail because of the different semantics of the `cd` command on Windows:  it doesn't change the default DRIVE if `cd` is executed from a different drive.  The result, when this is the case, is that Doxygen runs in the current working directory instead of in the temporary directory as was intended.  (fixed)
- when Doxygen exited with a non-zero result code, `sys.stdout.write()` was being called passing it a byte array instead of a string, and this caused it the Python script to generate an exception and abort.  (fixed)
- LVGL_URLPATH environment variable was being populated with a multi-line list of all current LVGL branches, rather than just the name of the current branch as intended.

The result:  the `gen_json.py` script now runs correctly on a Windows platform with TMPDIR environment variable set to a different drive than the script is executing from, when the local LVGL repository has multiple live branches.

**Comment by kisvegabor** [comment]
I've just tested it and it works on my end too. 

However when I changed only a single file it still recompiled everything. I'm using `./build.py skip_latex`.

I also noticed a larger amount of red text (warnings?)  Is it expected? 
See in [docs.txt](https://github.com/user-attachments/files/18029652/docs.txt)

**Comment by vwheeler63** [comment]
> I've just tested it and it works on my end too.

If you haven't already done so, you will need to follow the instructions in the module docstring, namely first:

- Docs-Dev Initial Docs Generation Usage 
    - Set `LVGL_FIXED_TEMP_DIR` environment variable to path to the temporary directory you will use over and over during         document editing, without trailing directory separator.  Initially directory should not exist.
    - $ python  build.py  skip_latex  preserve  fixed_tmp_dir
    - This takes about 22 minutes on first run.
- Docs-Dev Update-Only Generation Usage
    - $ python  build.py  skip_latex  docs_dev  update
    - Time to run depends on number of documents that were updated.

> I also noticed a larger amount of red text (warnings?) Is it expected?

Yes.  Some items are fixed in pending PRs.  Others I will target later (e.g. API `.rst` files).

- `./examples/widgets/scale/index.rst`   is fixed in a pending PR.
- API `.rst` files we can ignore for now.
- `./details/other-components/xml.rst`   I already fixed in a PR I will be submitting soon in next proofread batch.  You don't need to do anything.
- `./details/index.rst: WARNING: document isn't included in any toctree`    is expected.
- `./intro/index.rst: WARNING: document isn't included in any toctree`     is expected
- `./details/widgets/spangroup.rst: WARNING: document isn't included in any toctree`    is fixed in pending PR #7405 .
- `./details/main-components/display.rst:nnn:  ERROR: Error in "code-block" directive`   both errors are fixed in pending PR #7405 .\

Additional errors (if there are any) I will address later.  That is 99% of them.

**Comment by vwheeler63** [comment]
@liamHowatt @kisvegabor I haven't mentioned it yet, but I am also working TOWARDS getting the Latex => PDF build working again.  This involves removing Unicode from the files I visit.  In the PR I am about to submit (batch 12), 7 or 8 instances of Unicode embedded in the `font.rst` file has been removed and replaced with `replace` directives that keep the `.rst` file limited to the ASCII range and the Unicode characters are substituted during Sphinx processing.

I don't know how many docs have Unicode in them yet, but Kevin says that's the reason the Latex build won't work currently.

**Comment by vwheeler63** [comment]
@kisvegabor Did you try it again using the [instructions above](https://github.com/lvgl/lvgl/pull/7414#issuecomment-2521528911)?

**Comment by kisvegabor** [comment]
I'm doing this:
```
export LVGL_FIXED_TEMP_DIR="~/Desktop/tmp"
python build.py skip_latex preserve fixed_tmp_dir
python build.py skip_latex docs_dev update
```

But the update takes as much time as the normal run.

**Comment by vwheeler63** [comment]
> I'm doing this:
> 
> ```
> export LVGL_FIXED_TEMP_DIR="~/Desktop/tmp"
> python build.py skip_latex preserve fixed_tmp_dir
> python build.py skip_latex docs_dev update
> ```
> 
> But the update takes as much time as the normal run.

Can you share Sphinx's output to the screen?  Often the output gives a hint about why Sphinx is choosing to re-generate everything.

**Comment by vwheeler63** [comment]
@kisvegabor Do you have a Python debugger on your system?  I've been using PyCharm on Kevin's suggestion, and it has been working really nicely.  (I does look like JetBrains has a Linux version of PyCharm.)  I was thinking, if we needed to, we could share screens somehow and I could help you sort it out so Sphinx was not trying to rebuild everything.  Let me know.

Alternately, I could get my Ubuntu WSL going, although I have not done it yet, so I'm not sure how much time it would take to set up my Ubuntu system like yours.

**Comment by liamHowatt** [comment]
It worked for me!

(Working directory is `/home/liam/lvgl/lvgl/docs`)

```
LVGL_FIXED_TEMP_DIR=/home/liam/lvgl/doc_dev   python  build.py  skip_latex  preserve  fixed_tmp_dir
LVGL_FIXED_TEMP_DIR=/home/liam/lvgl/doc_dev   python  build.py  skip_latex  docs_dev  update
```

I can change something in `button.rst` and it rebuilds in < 10 seconds.

**Comment by vwheeler63** [comment]
@kisvegabor @liamHowatt 

Just to share a thought openly, a "normal" Sphinx doc-generation environment (sphinx-quickstart sets this up, and `python/cpython` , `writethedocs/www`, and `sphinx-doc/sphinx` all use it so it is safe to say it was the designers intent that it be this easy) uses this command line to build docs:
```
make html
# or
make latex
# or
make pdf
```

Since we do "certain things" to the source docs before the build takes place, we need a script to accomplish that (`build.py`), but I will be thinking as I go along about how to simplify the process in a way that is workable for all platforms.

**Comment by kisvegabor** [comment]
Sorry, I have a lot of things on my table now. If it worked for Liam, I'm sure that it's good.

**Comment by vwheeler63** [comment]
Hi, Liam!  ( @liamHowatt )

Question:  how do you get this onto your local system to try it?  Do you clone my fork and try it from that branch?  Or is there an easier way?

Is it continuing to work okay for you?

**Comment by liamHowatt** [comment]
Yes, I fetch your remote and checkout and pull this branch to use it. Actually, I use `gh co 7414`. (see [gh](https://cli.github.com/))

I've yet to take a proper look at how you did it, but as a black box, it works flawlessly for iterating on one RST file.

**Review by liamHowatt — COMMENTED** [review]
There are a lot of combinations of args and some args switch each other on and off. I think it's ok since you documented common arg combinations.

I wish we could know if it will break `master` before merging it. I'm inclined to approve if there's no solution for now.

**Comment by vwheeler63** [comment]
@liamHowatt I can confirm that all functionality is preserved for the original command line:
```
python  build.py  skip_latex
```

**Review by kisvegabor — APPROVED** [review]

**Review by liamHowatt — APPROVED** [review]
Okay, thanks!

**Comment by kdschlosser** [comment]
I did want to make a suggestion. I know this has already been merged but as a best practice the way sphinx is designed to work is everything should be running from the sphinx config file. build.py should not be something that gets directly run from the command line. To compile you should be able to `cd` into the docs directory and run "sphinx html" or "sphinx latex" and it is supposed to compile the documentation.

This is something that I have wanted to do but I never got around to doing it. 

@vwheeler63 
If you give me a call I can explain this in more detail as it is easier to do that than it is to type it all out.

**Comment by vwheeler63** [comment]
> I did want to make a suggestion. I know this has already been merged but as a best practice the way sphinx is designed to work is everything should be running from the sphinx config file. build.py should not be something that gets directly run from the command line. To compile you should be able to `cd` into the docs directory and run "sphinx html" or "sphinx latex" and it is supposed to compile the documentation.
> 
> This is something that I have wanted to do but I never got around to doing it.
> 
> @vwheeler63 If you give me a call I can explain this in more detail as it is easier to do that than it is to type it all out.

I think you mean:
```
make html
and
make latex
```
which then runs `sphinx-build`.  I have a dozen Sphinx projects on my system, and simplifying a docs-dev environment to those commands has been on my radar as well.

**Pull request merged**
